File: backend/app/services/calculation/cap_bank_ledger.py
"""
Cap bank ledger service.

Reconstructs the year-by-year cap bank timeline from finalized snapshots.
Surfaces the banked/drawn-down capacity that already exists in the cap
calculation engine but was previously invisible to users.
"""

import logging
from datetime import date, datetime
from decimal import ROUND_HALF_UP, Decimal
from typing import Any
from uuid import UUID

from app.services.calculation.caps import CapType
from app.services.calculation.models import (
    CapBankLedger,
    CapBankLedgerEntry,
)

logger = logging.getLogger(__name__)


def simulate_cap_bank(
    base_amount: Decimal,
    cap_rate: Decimal | None = None,
    cap_fixed_amount: Decimal | None = None,
    actual_amounts: list[Decimal] | None = None,
    cap_type: str = CapType.CUMULATIVE,
) -> list[CapBankLedgerEntry]:
    """Simulate year-by-year cap bank balances.

    Uses the same simulation logic as caps.py:295-314 but returns
    per-year entries with opening/closing bank balances.

    Args:
        base_amount: Base year amount (year 0).
        cap_rate: Annual cap rate (e.g., 0.05 for 5%).
        cap_fixed_amount: Fixed dollar max increase per year.
        actual_amounts: Ordered list of actual expenses per year.
        cap_type: Type of cap (cumulative or cumulative_compounding).

    Returns:
        List of CapBankLedgerEntry, one per year.
    """
    if actual_amounts is None or len(actual_amounts) == 0:
        return []

    if cap_type not in (CapType.CUMULATIVE, CapType.CUMULATIVE_COMPOUNDING):
        return []

    if cap_rate is None and cap_fixed_amount is None:
        return []

    q = Decimal("0.01")

    # Calculate annual increase limit
    if cap_type == CapType.CUMULATIVE:
        if cap_fixed_amount is not None:
            annual_increase_limit = cap_fixed_amount
        else:
            assert cap_rate is not None
            annual_increase_limit = (base_amount * cap_rate).quantize(q, ROUND_HALF_UP)
    else:
        # For compounding, annual_increase varies per year — computed in loop
        annual_increase_limit = None

    entries: list[CapBankLedgerEntry] = []
    running_reference = base_amount
    running_bank = Decimal("0")

    for year_idx, actual in enumerate(actual_amounts):
        bank_opening = running_bank.quantize(q, ROUND_HALF_UP)

        if cap_type == CapType.CUMULATIVE_COMPOUNDING:
            # Compounding: max = base * (1 + rate)^year
            years_since_base = year_idx + 1
            if cap_fixed_amount is not None:
                cap_threshold = (
                    base_amount + cap_fixed_amount * years_since_base
                ).quantize(q, ROUND_HALF_UP)
            else:
                assert cap_rate is not None
                cap_threshold = (
                    base_amount * ((Decimal("1") + cap_rate) ** years_since_base)
                ).quantize(q, ROUND_HALF_UP)

            # Effective max includes banked capacity
            effective_max = (cap_threshold + running_bank).quantize(q, ROUND_HALF_UP)
        else:
            # Cumulative (linear): reference + annual_increase + bank
            assert annual_increase_limit is not None
            cap_threshold = (running_reference + annual_increase_limit).quantize(
                q, ROUND_HALF_UP
            )
            effective_max = (cap_threshold + running_bank).quantize(q, ROUND_HALF_UP)

        # Apply cap
        if actual <= effective_max:
            amount_applied = actual
            excess = Decimal("0")
            new_bank = (effective_max - actual).quantize(q, ROUND_HALF_UP)
        else:
            amount_applied = effective_max
            excess = (actual - effective_max).quantize(q, ROUND_HALF_UP)
            new_bank = Decimal("0")

        bank_change = (new_bank - bank_opening).quantize(q, ROUND_HALF_UP)

        entries.append(
            CapBankLedgerEntry(
                period_start=date(2000 + year_idx, 1, 1),  # Placeholder dates
                period_end=date(2000 + year_idx, 12, 31),
                snapshot_id=None,
                cap_type=cap_type,
                cap_rate=cap_rate or Decimal("0"),
                base_year_amount=base_amount,
                cap_threshold=cap_threshold,
                actual_expense=actual,
                amount_applied=amount_applied,
                excess_absorbed_by_landlord=excess,
                bank_opening=bank_opening,
                bank_change=bank_change,
                bank_closing=new_bank,
                finalized_at=None,
            )
        )

        # Move reference forward for next year (cumulative linear only)
        if cap_type == CapType.CUMULATIVE:
            running_reference = actual
        running_bank = new_bank

    return entries


def get_cap_bank_ledger(
    lease_id: UUID,
    supabase_client: Any,
) -> CapBankLedger:
    """Build cap bank ledger from finalized reconciliation snapshots.

    Args:
        lease_id: The lease to build the ledger for.
        supabase_client: Supabase client for DB queries.

    Returns:
        CapBankLedger with all entries, or empty entries if no cumulative cap.

    Raises:
        ValueError: If lease not found.
    """
    # Fetch lease
    lease_response = (
        supabase_client.table("leases").select("*").eq("id", str(lease_id)).execute()
    )
    leases = lease_response.data if hasattr(lease_response, "data") else []
    if not leases:
        raise ValueError(f"Lease {lease_id} not found")

    lease = leases[0]
    recovery_profile = lease.get("recovery_profile", {})
    cap_type = recovery_profile.get("cap_type", "none")
    cap_rate_str = recovery_profile.get("cap_rate")
    cap_fixed_amount_str = recovery_profile.get("cap_fixed_amount")
    base_year_amount_str = recovery_profile.get("base_year_amount")
    tenant_name = lease.get("tenant_name", "")

    # If no cumulative cap, return empty ledger
    if cap_type not in (CapType.CUMULATIVE, CapType.CUMULATIVE_COMPOUNDING):
        return CapBankLedger(
            lease_id=lease_id,
            tenant_name=tenant_name,
            cap_type=cap_type,
            cap_rate=Decimal("0"),
            entries=[],
            current_bank_balance=Decimal("0"),
            total_landlord_absorbed=Decimal("0"),
        )

    cap_rate = Decimal(cap_rate_str) if cap_rate_str else None
    cap_fixed_amount = Decimal(cap_fixed_amount_str) if cap_fixed_amount_str else None
    base_year_amount = (
        Decimal(base_year_amount_str) if base_year_amount_str else Decimal("0")
    )

    # Fetch all finalized snapshots for this lease, ordered by period
    snapshots_response = (
        supabase_client.table("reconciliation_snapshots")
        .select("*")
        .eq("lease_id", str(lease_id))
        .eq("status", "finalized")
        .order("period_start_date", desc=False)
        .execute()
    )
    snapshots = snapshots_response.data if hasattr(snapshots_response, "data") else []

    if not snapshots:
        return CapBankLedger(
            lease_id=lease_id,
            tenant_name=tenant_name,
            cap_type=cap_type,
            cap_rate=cap_rate or Decimal("0"),
            entries=[],
            current_bank_balance=Decimal("0"),
            total_landlord_absorbed=Decimal("0"),
        )

    # Extract actual amounts from snapshots
    actual_amounts = [
        Decimal(str(s.get("tenant_share_before_cap", "0"))) for s in snapshots
    ]

    # Run bank simulation
    raw_entries = simulate_cap_bank(
        base_amount=base_year_amount,
        cap_rate=cap_rate,
        cap_fixed_amount=cap_fixed_amount,
        actual_amounts=actual_amounts,
        cap_type=cap_type,
    )

    # Enrich entries with actual snapshot metadata
    entries: list[CapBankLedgerEntry] = []
    for i, entry in enumerate(raw_entries):
        snapshot = snapshots[i]
        entries.append(
            entry.model_copy(
                update={
                    "period_start": _parse_date(snapshot.get("period_start_date", "")),
                    "period_end": _parse_date(snapshot.get("period_end_date", "")),
                    "snapshot_id": UUID(snapshot["id"]),
                    "finalized_at": _parse_datetime(snapshot.get("finalized_at")),
                }
            )
        )

    current_bank = entries[-1].bank_closing if entries else Decimal("0")
    total_absorbed = sum((e.excess_absorbed_by_landlord for e in entries), Decimal("0"))

    return CapBankLedger(
        lease_id=lease_id,
        tenant_name=tenant_name,
        cap_type=cap_type,
        cap_rate=cap_rate or Decimal("0"),
        entries=entries,
        current_bank_balance=current_bank,
        total_landlord_absorbed=total_absorbed,
    )


def _parse_date(date_str: str) -> date:
    """Parse date string to date object."""
    if not date_str:
        return date(2000, 1, 1)
    return date.fromisoformat(date_str[:10])


def _parse_datetime(dt_str: str | None) -> datetime | None:
    """Parse datetime string to datetime object."""
    if not dt_str:
        return None
    return datetime.fromisoformat(dt_str.replace("Z", "+00:00"))

File: cloudflare-backend/src/domain/reconciliation/cap-bank-ledger.ts
/**
 * Cap bank ledger domain logic.
 *
 * Faithful TypeScript port of backend/app/services/calculation/cap_bank_ledger.py.
 * All monetary arithmetic uses full-precision integer cent math (Money class) or
 * Decimal.js with precision=28, rounding ROUND_HALF_UP at each quantize step.
 *
 * CRITICAL: The compounding power base*(1+rate)^n is computed in full Decimal
 * precision and quantized ONCE to cents — matching Python's single-quantize.
 * Do NOT use the Rate type (8 dp) for this; it would diverge on multi-year runs.
 */

import Decimal from "decimal.js";

// ---------------------------------------------------------------------------
// Configure Decimal for this module: precision 28, ROUND_HALF_UP (3)
// ---------------------------------------------------------------------------
const D = Decimal.clone({ precision: 28, rounding: Decimal.ROUND_HALF_UP });

/** Quantize a Decimal to 2 decimal places with ROUND_HALF_UP */
function q(value: Decimal): Decimal {
  return value.toDecimalPlaces(2, Decimal.ROUND_HALF_UP);
}

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type CapBankLedgerEntry = {
  period_start: string; // YYYY-MM-DD
  period_end: string; // YYYY-MM-DD
  snapshot_id: string | null;
  cap_type: string;
  cap_rate: string; // serialized as string (Pydantic Decimal → JSON number, but frontend types.gen.ts says string)
  base_year_amount: string;
  cap_threshold: string;
  actual_expense: string;
  amount_applied: string;
  excess_absorbed_by_landlord: string;
  bank_opening: string;
  bank_change: string;
  bank_closing: string;
  finalized_at: string | null; // ISO datetime or null
};

export type CapBankLedger = {
  lease_id: string;
  tenant_name: string;
  pool_name: string | null;
  cap_type: string;
  cap_rate: string;
  entries: CapBankLedgerEntry[];
  current_bank_balance: string;
  total_landlord_absorbed: string;
};

export type SimulationInput = {
  baseAmount: string; // numeric string
  capRate: string | null; // numeric string or null
  capFixedAmount: string | null; // numeric string or null
  actualAmounts: string[]; // ordered list, one per finalized period
  capType: string;
};

export type SimulationEntry = Omit<
  CapBankLedgerEntry,
  "period_start" | "period_end" | "snapshot_id" | "finalized_at"
> & {
  /** placeholder; caller replaces with real snapshot data */
  period_start: string;
  period_end: string;
  snapshot_id: null;
  finalized_at: null;
};

// ---------------------------------------------------------------------------
// Cap type constants (mirror Python caps.py CapType)
// ---------------------------------------------------------------------------
const CAP_TYPE_CUMULATIVE = "cumulative";
const CAP_TYPE_COMPOUNDING = "cumulative_compounding";

// ---------------------------------------------------------------------------
// Core simulation — mirrors simulate_cap_bank exactly
// ---------------------------------------------------------------------------

export function simulateCapBank(input: SimulationInput): SimulationEntry[] {
  const { baseAmount, capRate, capFixedAmount, actualAmounts, capType } = input;

  if (actualAmounts.length === 0) {
    return [];
  }

  if (capType !== CAP_TYPE_CUMULATIVE && capType !== CAP_TYPE_COMPOUNDING) {
    return [];
  }

  if (capRate === null && capFixedAmount === null) {
    return [];
  }

  const base = new D(baseAmount);
  const rate = capRate !== null ? new D(capRate) : null;
  const fixedAmount = capFixedAmount !== null ? new D(capFixedAmount) : null;

  // For cumulative (linear), pre-compute annual_increase_limit
  let annualIncreaseLimit: Decimal | null = null;
  if (capType === CAP_TYPE_CUMULATIVE) {
    if (fixedAmount !== null) {
      annualIncreaseLimit = fixedAmount;
    } else {
      // cap_rate is not null here (guarded above)
      annualIncreaseLimit = q(base.mul(rate!));
    }
  }

  const entries: SimulationEntry[] = [];
  let runningReference = base;
  let runningBank = new D("0");

  for (let yearIdx = 0; yearIdx < actualAmounts.length; yearIdx++) {
    const actual = new D(actualAmounts[yearIdx]!);
    const bankOpening = q(runningBank);

    let capThreshold: Decimal;
    let effectiveMax: Decimal;

    if (capType === CAP_TYPE_COMPOUNDING) {
      const yearsSinceBase = yearIdx + 1;
      if (fixedAmount !== null) {
        // base + fixed * years
        capThreshold = q(base.add(fixedAmount.mul(yearsSinceBase)));
      } else {
        // base * (1 + rate)^years — full precision, quantize once
        const multiplier = new D("1").add(rate!).pow(yearsSinceBase);
        capThreshold = q(base.mul(multiplier));
      }
      effectiveMax = q(capThreshold.add(runningBank));
    } else {
      // Cumulative linear
      capThreshold = q(runningReference.add(annualIncreaseLimit!));
      effectiveMax = q(capThreshold.add(runningBank));
    }

    let amountApplied: Decimal;
    let excess: Decimal;
    let newBank: Decimal;

    if (actual.lte(effectiveMax)) {
      amountApplied = actual;
      excess = new D("0");
      newBank = q(effectiveMax.sub(actual));
    } else {
      amountApplied = effectiveMax;
      excess = q(actual.sub(effectiveMax));
      newBank = new D("0");
    }

    const bankChange = q(newBank.sub(bankOpening));

    entries.push({
      period_start: `${2000 + yearIdx}-01-01`,
      period_end: `${2000 + yearIdx}-12-31`,
      snapshot_id: null,
      cap_type: capType,
      cap_rate: (rate ?? new D("0")).toFixed(8).replace(/\.?0+$/, "") || "0",
      base_year_amount: base.toFixed(2),
      cap_threshold: capThreshold.toFixed(2),
      actual_expense: actual.toFixed(2),
      amount_applied: amountApplied.toFixed(2),
      excess_absorbed_by_landlord: excess.toFixed(2),
      bank_opening: bankOpening.toFixed(2),
      bank_change: bankChange.toFixed(2),
      bank_closing: newBank.toFixed(2),
      finalized_at: null,
    });

    // Carry forward: cumulative uses actual as next reference; compounding keeps base
    if (capType === CAP_TYPE_CUMULATIVE) {
      runningReference = actual;
    }
    runningBank = newBank;
  }

  return entries;
}

// ---------------------------------------------------------------------------
// Ledger builder — mirrors get_cap_bank_ledger
// ---------------------------------------------------------------------------

export type LeaseCapProfile = {
  leaseId: string;
  tenantName: string;
  capType: string;
  capRate: string | null;
  capFixedAmount: string | null;
  baseYearAmount: string | null;
};

export type FinalizedSnapshotRow = {
  id: string;
  tenant_share_before_cap: string | number | null;
  // Date/timestamp columns may arrive as `Date` if not `::text`-cast; parseDate
  // normalizes them. listFinalizedSnapshotsForLease casts them to text.
  period_start_date: string | Date;
  period_end_date: string | Date;
  finalized_at: string | Date | null;
};

export function buildCapBankLedger(
  profile: LeaseCapProfile,
  snapshots: FinalizedSnapshotRow[],
): CapBankLedger {
  const {
    leaseId,
    tenantName,
    capType,
    capRate,
    capFixedAmount,
    baseYearAmount,
  } = profile;

  const emptyLedger = (rate: string): CapBankLedger => ({
    lease_id: leaseId,
    tenant_name: tenantName,
    pool_name: null,
    cap_type: capType,
    cap_rate: rate,
    entries: [],
    current_bank_balance: "0.00",
    total_landlord_absorbed: "0.00",
  });

  if (capType !== CAP_TYPE_CUMULATIVE && capType !== CAP_TYPE_COMPOUNDING) {
    return emptyLedger("0");
  }

  const resolvedRate = capRate ?? null;
  const rateDisplay =
    resolvedRate !== null
      ? new D(resolvedRate).toFixed(8).replace(/\.?0+$/, "") || "0"
      : "0";

  if (snapshots.length === 0) {
    return emptyLedger(rateDisplay);
  }

  const actualAmounts = snapshots.map((s) =>
    String(s.tenant_share_before_cap ?? "0"),
  );

  const rawEntries = simulateCapBank({
    baseAmount: baseYearAmount ?? "0",
    capRate: resolvedRate,
    capFixedAmount: capFixedAmount ?? null,
    actualAmounts,
    capType,
  });

  // Overlay real snapshot metadata
  const entries: CapBankLedgerEntry[] = rawEntries.map((entry, i) => {
    const snapshot = snapshots[i]!;
    return {
      ...entry,
      period_start: parseDate(snapshot.period_start_date),
      period_end: parseDate(snapshot.period_end_date),
      snapshot_id: snapshot.id,
      finalized_at:
        snapshot.finalized_at instanceof Date
          ? snapshot.finalized_at.toISOString()
          : (snapshot.finalized_at ?? null),
    };
  });

  const currentBank =
    entries.length > 0 ? entries[entries.length - 1]!.bank_closing : "0.00";
  const totalAbsorbed = entries
    .reduce(
      (sum, e) => sum.add(new D(e.excess_absorbed_by_landlord)),
      new D("0"),
    )
    .toFixed(2);

  return {
    lease_id: leaseId,
    tenant_name: tenantName,
    pool_name: null,
    cap_type: capType,
    cap_rate: rateDisplay,
    entries,
    current_bank_balance: currentBank,
    total_landlord_absorbed: totalAbsorbed,
  };
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function parseDate(dateStr: string | Date | null | undefined): string {
  if (!dateStr) return "2000-01-01";
  // Date/timestamp columns decode to JS Date unless `::text`-cast; normalize so
  // a Date here does not throw `value.slice is not a function`.
  if (dateStr instanceof Date) {
    return dateStr.toISOString().slice(0, 10);
  }
  return dateStr.slice(0, 10);
}
